crates/pikelet-repl/src/lib.rs

```diff
@@ -176,86 +171,189 @@ pub fn run(opts: Opts) -> Result<(), Error> {
     Ok(())
 }
 
-fn eval_print(pikelet: &mut Driver, filemap: &FileMap) -> Result<ControlFlow, Vec<Diagnostic>> {
+#[derive(Clone)]
+enum ControlFlow {
+    Break,
+    Continue,
+}
+
+/// Commands entered in the REPL
+#[derive(Debug, Clone)]
+pub enum ReplCommand {
+    /// Evaluate a term
+    ///
+    /// ```text
+    /// <term>
+    /// ```
+    Eval(String),
+    /// Show the core representation of a term
+    ///
+    /// ```text
+    /// :core <term>
+    /// ```
+    Core(String),
+    /// Print some help about using the REPL
+    ///
+    /// ```text
+    /// :?
+    /// :h
+    /// :help
+    /// ```
+    Help,
+    /// Add a declaration to the REPL environment
+    ///
+    /// ```text
+    /// :let <name> = <term>
+    /// ```
+    Let(String, String),
+    ///  No command
+    NoOp,
+    /// Quit the REPL
+    ///
+    /// ```text
+    /// :q
+    /// :quit
+    /// ```
+    Quit,
+    /// Print the type of the term
+    ///
+    /// ```text
+    /// :t <term>
+    /// :type <term>
+    /// ```
+    TypeOf(String),
+}
+
+impl FromStr for ReplCommand {
+    type Err = Vec<Diagnostic>;
+
+    fn from_str(src: &str) -> Result<ReplCommand, Vec<Diagnostic>> {
+        use combine::char::*;
+        use combine::*;
+
+        let anys1 = || many1(any());
+        let spaces = || skip_many(space());
+        let spaces1 = || skip_many1(space());
+        let ident = || {
+            value(())
+                .with(letter())
+                .and(many::<String, _>(alpha_num()))
+                .map(|(hd, tl)| format!("{}{}", hd, tl))
+        };
+
+        let cmd = choice((
+            token(':').with(choice((
+                attempt(
+                    choice((
+                        attempt(string("help")),
+                        attempt(string("?")),
+                        attempt(string("h")),
+                    ))
+                    .map(|_| ReplCommand::Help),
+                ),
+                attempt(
+                    choice((attempt(string("quit")), attempt(string("q"))))
+                        .map(|_| ReplCommand::Quit),
+                ),
+                attempt(
+                    string("core")
+                        .with(spaces1())
+                        .with(anys1())
+                        .map(ReplCommand::Core),
+                ),
+                attempt(
+                    choice((attempt(string("type")), string("t")))
+                        .with(spaces1())
+                        .with(anys1())
+                        .map(ReplCommand::TypeOf),
+                ),
+                attempt(
+                    string("let")
+                        .with(spaces())
```

Comment:
spaces1?

Reply:
😱 you are right!!!

Reply:
I'm not used to working without a lexer it seems 😅